### ISSUE TITLE
Attract/Main video Museum Exhibit

### ISSUE THREAD

**Issue by lwshuck** [KenT2/pipresents-gapless#86]
I am implementing a couple of museum exhibits (Frontiers of Flight, Love Field, Dallas) using Raspberry Pis. My goal is to use PiPresents with two MP4 video tracks: 1) A one minute silent loop to “Attract” attention and ask the viewer to press a button (connected to GPIO) to take them to track 2) The “Main” video, and then return to the Attract loop. I can do the two pieces and I can see how to do this if I use an image in lieu of an Attract video, however, this will run 24-7 and I will use the Attract loop as somewhat of a screen saver. I have spent days on this without a solution. Can you give me some guidance please?

**Comment by KenT2** [comment]
You need a repeating mediashow with one track, the Attract Video. The mediashow has a child track which is the main video.

**Comment by lwshuck** [comment]
Thank you for your response. I wanted to respond earlier but I wanted a successful response. However, I've spent all day,with no success, trying to implement a child track. I am currently trying to keep it simple using up/down controls to trigger. I have obviously not figured this out yet. Under "Medialists" I have media.json and child.json. Under "Tracks in Selected Medislist" media.json shows Attract.mp4 & child.json shows Main.mp4; and yet, Validate (& when I run it) I am told that Child not found in medialist:  child. What else do I need to do?

**Comment by KenT2** [comment]
The child track in the medialist must have a track reference
That reference must be in the Child Tack field of the Child Track tab of the mediashow
use the play command to play the child track

**Comment by lwshuck** [comment]
Thanks for your help. It is working very well with a keyboard and mouse. This is a significant step for me. I will now change to a GPIO trigger and Start on power-up. I've seen examples related to these two steps so I have high hopes. Thanks for PiPresents and your help.

**Comment by lwshuck** [comment]
As an old H/W guy (I designed RADAR systems) I am very pleased to have gotten our video exhibits working. Your Pi Presents has  been a huge help. I have one more issue that may not be a show stopper but it is at least an annoyance.
When the play button is pressed to switch from the looping Attract.mp4 to the Main.mp4 there is a one second flash (or delay). After re-booting the Desktop is displayed during this flash. This persists for ~10 minutes, after which there is a black screen during this flash. If it stayed like that I would be pleased. However, about every 3-4 hours it reverts to flashing the Desktop for another 10 minutes. It doesn't seem to be related to the number of times the play button is pressed. If you have a suggested fix, it would be greatly appreciated.
Thanks

**Comment by KenT2** [comment]
I think you need to add -b and -f command line options to disable blanking and set PP to full screen.

**Comment by lwshuck** [comment]
The -f solved my problem. I am ready to present our design to our curator and director. I will continue, as I have in my periodic status reports, to give attribution for my success to Pi Presents. This is the culmination of something I have been working on for two years.
I thank you greatly for your help.

**Comment by srauco** [comment]
How do you get back to the Attract without hitting escape? I want one video to play in a loop and when a button is pressed, it's interrupted by the child video. When the child finishes it goes back to the Attract. Can you possibly share your profile?

**Comment by KenT2** [comment]
Make the child mediashow single run

**Comment by srauco** [comment]
Thanks Ken, that worked well. I have almost everything going but I have one more question if you don't mind. I want to trigger the child show from a button on a gpio for PIR. Do I need to set the child as input? I have tried numerous ways using pins 9 and 11 and can't get it to trigger. I'm using the standard gpio text file in the templates.

**Comment by srauco** [comment]
Oh, I should probably mention. Without the gpio text file, and the child set to start and not input, the return key does trigger the child correctly

**Comment by Alchemist45** [comment]
Ken,

Can you ask lwshuck if willing to post the profile that finally worked? 

Thanks,
Scott

**Comment by KenT2** [comment]
Github does not allow me to contact @lwshuck hopefully he will see this exchange.
I think the example pp_interactive_1p3 is a good starting point for what you want to do.

**Comment by Alchemist45** [comment]
Hi Ken,

I am getting the following errors when I try to run.  Can you suggest what I'm doing wrong?

python /home/pi/pipresents/pipresents.py -p OurRegion_new -f -b
Traceback (most recent call last):
  File "/home/pi/pipresents/pipresents.py", line 860, in <module>
    pp = PiPresents()
  File "/home/pi/pipresents/pipresents.py", line 447, in __init__
    self.run_start_shows()           
  File "/home/pi/pipresents/pipresents.py", line 480, in run_start_shows
    reason,message=self.show_manager.control_a_show(show_ref,'open')
  File "/home/pi/pipresents/pp_showmanager.py", line 128, in control_a_show
    return self.start_show(show_ref)
  File "/home/pi/pipresents/pp_showmanager.py", line 175, in start_show
    show_obj.play(self._end_play_show,None,False,0,[])
  File "/home/pi/pipresents/pp_mediashow.py", line 36, in play
    GapShow.play(self,end_callback,show_ready_callback, direction_command,level,controls_list)
  File "/home/pi/pipresents/pp_gapshow.py", line 97, in play
    self.start_show()
  File "/home/pi/pipresents/pp_gapshow.py", line 237, in start_show
    self.wait_for_trigger()
  File "/home/pi/pipresents/pp_gapshow.py", line 265, in wait_for_trigger
    self.start_list()
  File "/home/pi/pipresents/pp_gapshow.py", line 302, in start_list
    self.what_next_after_showing()
  File "/home/pi/pipresents/pp_gapshow.py", line 543, in what_next_after_showing
    self.start_list()               
  File "/home/pi/pipresents/pp_gapshow.py", line 309, in start_list
    self.start_load_show_loop(self.medialist.selected_track())
  File "/home/pi/pipresents/pp_medialist.py", line 98, in selected_track
    return self._selected_track
AttributeError: 'MediaList' object has no attribute '_selected_track'
pi@OIRMPI1:~ $ unclutter: someone created a sub-window to my sub-window! giving up

**Comment by KenT2** [comment]
Looks like you have zero anonymous (blank track reference) tracks in the medialist. Does this problem show up if you validate the profile using the editor.

Think you might have discovered an uncaught profile error.

**Comment by lwshuck** [comment]
>I'm an old H/W guy so I'm not sure I have posted this the way you expected, but I believe the attached FOFM.zip file is what you are calling the profile I'm Using. This is what I have in my /home/pi/pp_home/pp_profiles directory.
[FOFM.zip](https://github.com/KenT2/pipresents-gapless/files/2674433/FOFM.zip)
***** New issue*****
I saw the following statement in the Manual:
"Pi Presents crashes after playing videos for a few hours
There seems to be some timing problem in the interface with OMXPlayer or with
OMXPlayer itself."

We have two Frontiers of Flight Museum (FOFM) displays that run 24/7. So before installing at the museum I tested my two units continuously for three weeks  with no issues. I installed the two units on Oct. 11 and one has now crashed. It appears to have locked up either at the beginning or the end of the one minute looping attract.mp4 file. Leaving a still image on the TV. Two questions:
1. Is there any new information on these OMXPlayer related crashes?
2. If I enable the scheduler does that reboot the program and thereby interrupting/reseting these long play times?

**Comment by KenT2** [comment]
I looked at your profile. You have seamless loop turned on in the attract track, this is not necessary as it is the only track in the show. You could try turning this off and ensure 'pause at end' is on to disguise the small gap while PP reloads the video . With modern PI's the load time is so small that seamless loop is only for the most demanding applications.
With seamless loop on the video track actually never ends because omxplayer does the looping so it sounds like an omxplayer issue, not a PP issue.

omxplayer bug reports are at https://github.com/popcornmix/omxplayer   You will see there are continuing issues with erratic behaviour. I get the occasional report of PP problems with video, they seem to be from people who have made their own videos, it might be worth re-rendering the video using a different package e.g. handbrake.

Other things to check:
GPU Memory is large enough
Swap power supplies, does the problem follow the power supply.

The scheduler does not reboot the program, it just opens and closes shows and will not sort out a crash. You could use something like cron to do this, it would need to ensure PP, omxplayer and omxplayer.bin are all closed.

**Comment by lwshuck** [comment]
Thank you for your response. I appreciate your suggestions. 

I have no idea whether this is helpful, but I have a friend who did his PhD EE dissertation in image processing. I have consulted with him on a very limited basis because he owns his own business and has very little personal time. He gave me a possible explanation which I find plausible. I will paraphrase, using my limited knowledge, in hopes that this may be helpful. I would I value your opinion.
_As a part of a memory management process memory is sometimes borrowed from here and there at particular times and then repaid later. However, there can be circumstances where the process fails to repay a few bits here and there and over a long period of time the management process runs out of usable memory and locks up. He suggested I research the possibility of using a watchdog timer that would do a reboot if the processor failed to reset the timer in a specified time frame.__

When I Googled “raspberry pi watchdog reboot” I was surprised by the number hits I received. I am currently putting a hard press on finishing a book that I am writing. Thus, it may be a few months before I allocate more time to this issue. But I’m raising this possible cause/solution for what it might be worth.

**Comment by lwshuck** [comment]
Without a S/W background I'm a very slow learner. I am currently researching "memory leak". When I google "memory leak in RPi", there's an abundance of info, e.g. https://www.raspberrypi.org/forums/viewtopic.php?t=16850. I don't physically go to the museum very often, so I'm evaluating a, periodic,  middle of the night reboot. If you find a more elegant solution I would appreciate hearing of it. Thanks.

**Comment by KenT2** [comment]
closed for now. Please re-open if you have a continuing problem

**Comment by lwshuck** [comment]
I implemented your suggested changes. They ran fine but at one time or another each of our two museum exhibits have continued to lock up. Two months seems to be typical. When I reboot all is well for another couple of months. I chose to implement a crontab file as follows:
1. At the terminal screen I ran "crontab -e"
2. I selected option 2 which is the NANO editor
3. The inputs are m=Minute, h=Hour, dom=Day of month, mon=Month of year, dow=Day of week, command  
4. I entered "0 24 * * 0 reboot"
5. I had difficulty figuring out how to save and exit, until I learned that Write Out means save (OK not that tuff) and that ^O = Ctrl+O. (I guess I will eventually learn how to speak S/W)
6. After saving exiting and rebooting crontab executes a reboot each Sunday at midnight.
7. I only recently installed the changes at the museum, testing here at home looked good. Only time will tell.
I was concerned about the RPi clock staying sync'd. But it appears that the RPi remains sync'd, using an NTP server as long as it's on WiFi and has access to the internet (which I have).

**Comment by KenT2** [comment]
I suspect that the problem is not that PP is failing after 2 months but that the time to failure is random so in theory it might fail within hours if you are unlucky.

Would it be better if you do the reboot every night then if the PI has failed the day before it will be up and working for the next day. Most museums seem to turn off all their exhibits at night anyway for fire safety.

Have a look at this issue:
[https://github.com/KenT2/pipresents-beep/issues/3](url)

We came to the conclusion that the problem here was serving videos over a network. Are you doing that?

**Comment by lwshuck** [comment]
If the lock-up is not time-related, i.e. memory bleed, then I’m not sure reboot is the answer. I put two buttons on the cover of our RPi’s, “Play” & “Shut Down”. “Play” is in parallel with the button around on the front of the display and allows me to test from back behind. It is probably overkill, but after reading about the risks associated with simply pulling the power I decided to implement a button that uses GPIO-18 (P1-12) to perform a graceful shutdown. When ours locks-up it freezes while displaying a frame of the “Attract” video. At that point, it will not respond to my shutdown button. I’ve found I must kill the power to reboot. Thus, I don’t believe Cron will ever run.

**Comment by KenT2** [comment]
P1-12 will not always work  as its response is by  Pi Presents which is itself locked up in a way that stops it testing the GPIO line. Cron is an independent program and should work.
It would be useful to see the log file from when your video failed. If you use the -d command line option the log is saved to the file /pp_logs/pp_log.txt. Either post it here or contact me via the wordpress blog to get my email.

**Comment by lwshuck** [comment]
Thanks,
I am currently running two PP displays at the museum that reboot nightly. I am running a third one at home with the S/W that locked up at the museum (-d command added). If/when it locks up I will send you the /pp_logs/pp_log.txt file. Currently at 16 days and running.

**Comment by lwshuck** [comment]
I have experienced a failure of the PP that I'm running at home w/o the nightly reboot. This failure is very different from previous. On previous occasions the screen locked with a still frame from the attract.mp4. This time It was locked up w/ a blank screen. I do not have enough expertise to glean anything from the attached log but I hope you find it useful. Thanks

[pp_log.txt](https://github.com/KenT2/pipresents-gapless/files/3343238/pp_log.txt)

**Comment by lwshuck** [comment]
In trying to analyze the pp_log.txt, it appears the presentation was terminated by the GPIO shut down button I have implemented and not an unexpected failure. This is consistent with the black screen I observed. How this happened I'm not sure. I suppose someone (most likely me) laid something on my RPi and shut it down.
To verify this assumption I shut down and restarted, which produced a pp_log consistent the attached pp_log.txt. Please excuse me for my false alarm.
I have re-started the test unit and will continue to try to collect a pp_ log from a failed run.

**Comment by KenT2** [comment]
Had a look at the log. I think you sent the wrong one.  I have deinterleaved the tracks in the attached log and it shows that the last track was terminated by closing the window before the track was due to complete.
[lwshuck pp_log.txt](https://github.com/KenT2/pipresents-gapless/files/3356593/lwshuck.pp_log.txt)
